// @remotion/media-parser/dist/containers/m3u/iterate-over-segment-files.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.iteratorOverSegmentFiles = void 0;
const media_parser_controller_1 = require("../../controller/media-parser-controller");
const forward_controller_pause_resume_abort_1 = require("../../forward-controller-pause-resume-abort");
const parse_media_1 = require("../../parse-media");
const with_resolvers_1 = require("../../with-resolvers");
const get_chunks_1 = require("./get-chunks");
const get_playlist_1 = require("./get-playlist");
const iteratorOverSegmentFiles = async ({ structure, onVideoTrack, m3uState, onAudioTrack, onDoneWithTracks, playlistUrl, logLevel, parentController, onInitialProgress, readerInterface, }) => {
    const playlist = (0, get_playlist_1.getPlaylist)(structure, playlistUrl);
    const chunks = (0, get_chunks_1.getChunks)(playlist);
    let resolver = onInitialProgress;
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    let rejector = (_e) => { };
    for (const chunk of chunks) {
        resolver = onInitialProgress;
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        rejector = (_e) => { };
        const childController = (0, media_parser_controller_1.mediaParserController)();
        const forwarded = (0, forward_controller_pause_resume_abort_1.forwardMediaParserControllerPauseResume)({
            childController,
            parentController,
        });
        const makeContinuationFn = () => {
            return {
                continue() {
                    const { promise, reject, resolve } = (0, with_resolvers_1.withResolvers)();
                    resolver = resolve;
                    rejector = reject;
                    childController.resume();
                    return promise;
                },
                abort() {
                    childController.abort();
                },
            };
        };
        const isLastChunk = chunk === chunks[chunks.length - 1];
        await childController._internals.checkForAbortAndPause();
        const src = readerInterface.createAdjacentFileSource(chunk.url, playlistUrl);
        try {
            const mp4HeaderSegment = m3uState.getMp4HeaderSegment(playlistUrl);
            const data = await (0, parse_media_1.parseMedia)({
                src,
                acknowledgeRemotionLicense: true,
                logLevel,
                controller: childController,
                progressIntervalInMs: 0,
                onParseProgress: () => {
                    childController.pause();
                    resolver(makeContinuationFn());
                },
                fields: chunk.isHeader ? { structure: true } : undefined,
                onTracks: () => {
                    if (!m3uState.hasEmittedDoneWithTracks(playlistUrl)) {
                        m3uState.setHasEmittedDoneWithTracks(playlistUrl);
                        onDoneWithTracks();
                        return null;
                    }
                },
                onAudioTrack: onAudioTrack === null
                    ? null
                    : async ({ track }) => {
                        const callbackOrFalse = m3uState.hasEmittedAudioTrack(playlistUrl);
                        if (callbackOrFalse === false) {
                            const callback = await onAudioTrack(track);
                            if (!callback) {
                                m3uState.setHasEmittedAudioTrack(playlistUrl, null);
                                return null;
                            }
                            m3uState.setHasEmittedAudioTrack(playlistUrl, callback);
                            return (sample) => {
                                return callback(sample);
                            };
                        }
                        return callbackOrFalse;
                    },
                onVideoTrack: onVideoTrack === null
                    ? null
                    : async ({ track }) => {
                        const callbackOrFalse = m3uState.hasEmittedVideoTrack(playlistUrl);
                        if (callbackOrFalse === false) {
                            const callback = await onVideoTrack({
                                ...track,
                                m3uStreamFormat: chunk.isHeader || mp4HeaderSegment ? 'mp4' : 'ts',
                            });
                            if (!callback) {
                                m3uState.setHasEmittedVideoTrack(playlistUrl, null);
                                return null;
                            }
                            m3uState.setHasEmittedVideoTrack(playlistUrl, callback);
                            return (sample) => {
                                return callback(sample);
                            };
                        }
                        return callbackOrFalse;
                    },
                reader: readerInterface,
                mp4HeaderSegment,
                makeSamplesStartAtZero: false,
            });
            if (chunk.isHeader) {
                if (data.structure.type !== 'iso-base-media') {
                    throw new Error('Expected an mp4 file');
                }
                m3uState.setMp4HeaderSegment(playlistUrl, data.structure);
            }
        }
        catch (e) {
            rejector(e);
            throw e;
        }
        forwarded.cleanup();
        if (!isLastChunk) {
            childController.pause();
            resolver(makeContinuationFn());
        }
    }
    resolver(null);
};
exports.iteratorOverSegmentFiles = iteratorOverSegmentFiles;
