// @remotion/media-parser/dist/containers/m3u/run-over-m3u.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.runOverM3u = void 0;
const log_1 = require("../../log");
const register_track_1 = require("../../register-track");
const iterate_over_segment_files_1 = require("./iterate-over-segment-files");
const runOverM3u = async ({ state, structure, playlistUrl, logLevel, }) => {
    const tracksDone = state.m3u.getTrackDone(playlistUrl);
    const hasAudioStreamToConsider = state.m3u.sampleSorter.hasAudioStreamToConsider(playlistUrl);
    const hasVideoStreamToConsider = state.m3u.sampleSorter.hasVideoStreamToConsider(playlistUrl);
    const audioDone = !hasAudioStreamToConsider && tracksDone;
    const videoDone = !hasVideoStreamToConsider && tracksDone;
    const bothDone = audioDone && videoDone;
    if (bothDone) {
        state.m3u.setAllChunksProcessed(playlistUrl);
        return;
    }
    const existingRun = state.m3u.getM3uStreamRun(playlistUrl);
    if (existingRun) {
        log_1.Log.trace(logLevel, 'Existing M3U parsing process found for', playlistUrl);
        const run = await existingRun.continue();
        state.m3u.setM3uStreamRun(playlistUrl, run);
        if (!run) {
            state.m3u.setAllChunksProcessed(playlistUrl);
        }
        return;
    }
    log_1.Log.trace(logLevel, 'Starting new M3U parsing process for', playlistUrl);
    return new Promise((resolve, reject) => {
        const run = (0, iterate_over_segment_files_1.iteratorOverSegmentFiles)({
            playlistUrl,
            structure,
            onInitialProgress: (newRun) => {
                state.m3u.setM3uStreamRun(playlistUrl, newRun);
                resolve();
            },
            logLevel: state.logLevel,
            onDoneWithTracks() {
                const allDone = state.m3u.setTracksDone(playlistUrl);
                if (allDone) {
                    state.callbacks.tracks.setIsDone(state.logLevel);
                }
            },
            onAudioTrack: audioDone
                ? null
                : async (track) => {
                    const existingTracks = state.callbacks.tracks.getTracks();
                    let { trackId } = track;
                    while (existingTracks.find((t) => t.trackId === trackId)) {
                        trackId++;
                    }
                    const onAudioSample = await (0, register_track_1.registerAudioTrack)({
                        container: 'm3u8',
                        track: {
                            ...track,
                            trackId,
                        },
                        registerAudioSampleCallback: state.callbacks.registerAudioSampleCallback,
                        tracks: state.callbacks.tracks,
                        logLevel: state.logLevel,
                        onAudioTrack: state.onAudioTrack,
                    });
                    state.m3u.sampleSorter.addToStreamWithTrack(playlistUrl);
                    if (onAudioSample === null) {
                        return null;
                    }
                    state.m3u.sampleSorter.addAudioStreamToConsider(playlistUrl, onAudioSample);
                    return async (sample) => {
                        await state.m3u.sampleSorter.addAudioSample(playlistUrl, sample);
                    };
                },
            onVideoTrack: videoDone
                ? null
                : async (track) => {
                    const existingTracks = state.callbacks.tracks.getTracks();
                    let { trackId } = track;
                    while (existingTracks.find((t) => t.trackId === trackId)) {
                        trackId++;
                    }
                    const onVideoSample = await (0, register_track_1.registerVideoTrack)({
                        container: 'm3u8',
                        track: {
                            ...track,
                            trackId,
                        },
                        logLevel: state.logLevel,
                        onVideoTrack: state.onVideoTrack,
                        registerVideoSampleCallback: state.callbacks.registerVideoSampleCallback,
                        tracks: state.callbacks.tracks,
                    });
                    state.m3u.sampleSorter.addToStreamWithTrack(playlistUrl);
                    if (onVideoSample === null) {
                        return null;
                    }
                    state.m3u.sampleSorter.addVideoStreamToConsider(playlistUrl, onVideoSample);
                    return async (sample) => {
                        await state.m3u.sampleSorter.addVideoSample(playlistUrl, sample);
                    };
                },
            m3uState: state.m3u,
            parentController: state.controller,
            readerInterface: state.readerInterface,
        });
        run.catch((err) => {
            reject(err);
        });
    });
};
exports.runOverM3u = runOverM3u;
