// @remotion/media-parser/dist/readers/from-web-file.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.webFileReader = exports.webFileCreateAdjacentFileSource = exports.webFileReadWholeAsText = exports.webFileReadContent = void 0;
const webFileReadContent = ({ src, range, controller }) => {
    if (typeof src === 'string' || src instanceof URL) {
        throw new Error('`inputTypeFileReader` only supports `File` objects');
    }
    const part = range === null
        ? src
        : typeof range === 'number'
            ? src.slice(range)
            : src.slice(range[0], range[1]);
    const reader = new FileReader();
    reader.readAsArrayBuffer(src);
    const ownController = new AbortController();
    if (ownController) {
        ownController.signal.addEventListener('abort', () => {
            reader.abort();
        }, { once: true });
    }
    if (controller) {
        controller._internals.signal.addEventListener('abort', () => {
            ownController.abort();
        }, { once: true });
    }
    return new Promise((resolve, reject) => {
        reader.onload = () => {
            const stream = part.stream();
            const streamReader = stream.getReader();
            resolve({
                reader: {
                    reader: streamReader,
                    abort() {
                        streamReader.cancel();
                        ownController.abort();
                    },
                },
                contentLength: src.size,
                name: src instanceof File ? src.name : src.toString(),
                supportsContentRange: true,
                contentType: src.type,
                needsContentRange: true,
            });
        };
        reader.onerror = () => {
            reject(reader.error);
        };
    });
};
exports.webFileReadContent = webFileReadContent;
const webFileReadWholeAsText = () => {
    throw new Error('`webFileReader` cannot read auxiliary files.');
};
exports.webFileReadWholeAsText = webFileReadWholeAsText;
const webFileCreateAdjacentFileSource = () => {
    throw new Error('`webFileReader` cannot create adjacent file sources.');
};
exports.webFileCreateAdjacentFileSource = webFileCreateAdjacentFileSource;
exports.webFileReader = {
    read: exports.webFileReadContent,
    readWholeAsText: exports.webFileReadWholeAsText,
    createAdjacentFileSource: exports.webFileCreateAdjacentFileSource,
};
