// @remotion/media-parser/dist/containers/mp3/audio-sample-from-cbr.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAudioSampleFromCbr = void 0;
const get_frame_length_1 = require("./get-frame-length");
const getAudioSampleFromCbr = ({ bitrateInKbit, initialOffset, layer, sampleRate, samplesPerFrame, data, state, }) => {
    const avgLength = (0, get_frame_length_1.getAverageMpegFrameLength)({
        bitrateKbit: bitrateInKbit,
        layer,
        samplesPerFrame,
        samplingFrequency: sampleRate,
    });
    const mp3Info = state.mp3.getMp3Info();
    if (!mp3Info) {
        throw new Error('No MP3 info');
    }
    const nthFrame = Math.round((initialOffset - state.mediaSection.getMediaSectionAssertOnlyOne().start) /
        avgLength);
    const durationInSeconds = samplesPerFrame / sampleRate;
    const timeInSeconds = (nthFrame * samplesPerFrame) / sampleRate;
    const timestamp = Math.round(timeInSeconds * 1000000);
    const duration = Math.round(durationInSeconds * 1000000);
    const audioSample = {
        data,
        cts: timestamp,
        dts: timestamp,
        duration,
        offset: initialOffset,
        timescale: 1000000,
        timestamp,
        trackId: 0,
        type: 'key',
    };
    return audioSample;
};
exports.getAudioSampleFromCbr = getAudioSampleFromCbr;
