// @remotion/media-parser/dist/containers/iso-base-media/get-seeking-byte.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSeekingByteFromIsoBaseMedia = void 0;
const get_tracks_1 = require("../../get-tracks");
const log_1 = require("../../log");
const video_section_1 = require("../../state/video-section");
const collect_sample_positions_from_moof_boxes_1 = require("./collect-sample-positions-from-moof-boxes");
const find_keyframe_before_time_1 = require("./find-keyframe-before-time");
const get_sample_position_bounds_1 = require("./get-sample-position-bounds");
const get_sample_positions_from_track_1 = require("./get-sample-positions-from-track");
const find_best_segment_from_tfra_1 = require("./mfra/find-best-segment-from-tfra");
const traversal_1 = require("./traversal");
const getSeekingByteFromIsoBaseMedia = async ({ info, time, logLevel, currentPosition, isoState, mp4HeaderSegment, structure, }) => {
    const tracks = (0, get_tracks_1.getTracksFromMoovBox)(info.moovBox);
    const allTracks = [
        ...tracks.videoTracks,
        ...tracks.audioTracks,
        ...tracks.otherTracks,
    ];
    const hasMoov = Boolean((0, traversal_1.getMoovBoxFromState)({
        mp4HeaderSegment,
        structureState: structure,
        isoState,
        mayUsePrecomputed: false,
    }));
    if (!hasMoov) {
        log_1.Log.trace(logLevel, 'No moov box found, must wait');
        return {
            type: 'valid-but-must-wait',
        };
    }
    const firstVideoTrack = allTracks.find((t) => t.type === 'video');
    if (!firstVideoTrack) {
        throw new Error('No video track found');
    }
    const { timescale } = firstVideoTrack;
    if (info.moofBoxes.length > 0) {
        const tkhdBox = (0, traversal_1.getTkhdBox)(firstVideoTrack.trakBox);
        if (!tkhdBox) {
            throw new Error('Expected tkhd box in trak box');
        }
        const { samplePositions: samplePositionsArray } = (0, collect_sample_positions_from_moof_boxes_1.collectSamplePositionsFromMoofBoxes)({
            moofBoxes: info.moofBoxes,
            tfraBoxes: info.tfraBoxes,
            tkhdBox,
        });
        log_1.Log.trace(logLevel, 'Fragmented MP4 - Checking if we have seeking info for this time range');
        for (const positions of samplePositionsArray) {
            const { min, max } = (0, get_sample_position_bounds_1.getSamplePositionBounds)(positions, timescale);
            if (min <= time && time <= max) {
                log_1.Log.trace(logLevel, `Fragmented MP4 - Found that we have seeking info for this time range: ${min} <= ${time} <= ${max}`);
                const kf = (0, find_keyframe_before_time_1.findKeyframeBeforeTime)({
                    samplePositions: positions,
                    time,
                    timescale,
                    logLevel,
                    mediaSections: info.mediaSections,
                });
                if (kf) {
                    return {
                        type: 'do-seek',
                        byte: kf,
                    };
                }
            }
        }
        const atom = await (info.mfraAlreadyLoaded
            ? Promise.resolve(info.mfraAlreadyLoaded)
            : isoState.mfra.triggerLoad());
        if (atom) {
            const moofOffset = (0, find_best_segment_from_tfra_1.findBestSegmentFromTfra)({
                mfra: atom,
                time,
                firstVideoTrack,
                timescale,
            });
            if (moofOffset !== null &&
                !(moofOffset.start <= currentPosition &&
                    currentPosition < moofOffset.end)) {
                log_1.Log.verbose(logLevel, `Fragmented MP4 - Found based on mfra information that we should seek to: ${moofOffset.start} ${moofOffset.end}`);
                return {
                    type: 'intermediary-seek',
                    byte: moofOffset.start,
                };
            }
        }
        log_1.Log.trace(logLevel, 'Fragmented MP4 - No seeking info found for this time range.');
        if ((0, video_section_1.isByteInMediaSection)({
            position: currentPosition,
            mediaSections: info.mediaSections,
        }) !== 'in-section') {
            return {
                type: 'valid-but-must-wait',
            };
        }
        log_1.Log.trace(logLevel, 'Fragmented MP4 - Inside the wrong video section, skipping to the end of the section');
        const mediaSection = (0, video_section_1.getCurrentMediaSection)({
            offset: currentPosition,
            mediaSections: info.mediaSections,
        });
        if (!mediaSection) {
            throw new Error('No video section defined');
        }
        return {
            type: 'intermediary-seek',
            byte: mediaSection.start + mediaSection.size,
        };
    }
    const { samplePositions, isComplete } = (0, get_sample_positions_from_track_1.getSamplePositionsFromTrack)({
        trakBox: firstVideoTrack.trakBox,
        moofBoxes: info.moofBoxes,
        tfraBoxes: info.tfraBoxes,
    });
    if (!isComplete) {
        throw new Error('Incomplete sample positions');
    }
    const keyframe = (0, find_keyframe_before_time_1.findKeyframeBeforeTime)({
        samplePositions,
        time,
        timescale,
        logLevel,
        mediaSections: info.mediaSections,
    });
    if (keyframe) {
        return {
            type: 'do-seek',
            byte: keyframe,
        };
    }
    return {
        type: 'invalid',
    };
};
exports.getSeekingByteFromIsoBaseMedia = getSeekingByteFromIsoBaseMedia;
