// @remotion/media-parser/dist/state/riff/sample-counter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.riffSampleCounter = void 0;
const riff_keyframes_1 = require("./riff-keyframes");
const riffSampleCounter = () => {
    const samplesForTrack = {};
    const riffKeys = (0, riff_keyframes_1.riffKeyframesState)();
    const onAudioSample = (trackId, audioSample) => {
        if (typeof samplesForTrack[trackId] === 'undefined') {
            samplesForTrack[trackId] = 0;
        }
        if (audioSample.data.length > 0) {
            samplesForTrack[trackId]++;
        }
        samplesForTrack[trackId]++;
    };
    const onVideoSample = (trackId, videoSample) => {
        if (typeof samplesForTrack[trackId] === 'undefined') {
            samplesForTrack[trackId] = 0;
        }
        if (videoSample.type === 'key') {
            riffKeys.addKeyframe({
                trackId,
                decodingTimeInSeconds: videoSample.dts / videoSample.timescale,
                positionInBytes: videoSample.offset,
                presentationTimeInSeconds: videoSample.cts / videoSample.timescale,
                sizeInBytes: videoSample.data.length,
                sampleCounts: { ...samplesForTrack },
            });
        }
        if (videoSample.data.length > 0) {
            samplesForTrack[trackId]++;
        }
    };
    const getSamplesForTrack = (trackId) => {
        var _a;
        return (_a = samplesForTrack[trackId]) !== null && _a !== void 0 ? _a : 0;
    };
    const setSamplesFromSeek = (samples) => {
        for (const trackId in samples) {
            samplesForTrack[trackId] = samples[trackId];
        }
    };
    return {
        onAudioSample,
        onVideoSample,
        getSamplesForTrack,
        setSamplesFromSeek,
        riffKeys,
    };
};
exports.riffSampleCounter = riffSampleCounter;
